File: advall-python/d7.py
def isAbba(s):
    return (s[0] == s[3]) and (s[1] == s[2]) and (s[0] != s[1])

def containsAbba(s):
    if len(s) < 4:
        return False
    hasAbba = False
    for i in range(len(s) - 3):
        if isAbba(s[i:(i+4)]):
            hasAbba = True
            break
    return hasAbba

def isAba(s):
    if (s[0] == s[2]) and (s[1] != s[0]):
        return (s[0], s[1])
    else:
        return ('_', '_')


numOfTLS = 0
numOfSSL = 0
for line in open ("d7_input.txt", "r"):
    ip = line.rstrip()
    if ip[0] == '[':
        ip = "_" + ip
    ip = ip.replace(']', '[')
    ip = ip.split('[')
    isTLS = False
    for i in range(0, len(ip), 2):
        if containsAbba(ip[i]):
            isTLS = True
    for i in range(1, len(ip), 2):
        if containsAbba(ip[i]):
            isTLS = False
    if isTLS:
        numOfTLS += 1

    isSSL = False
    for i in range(0, len(ip), 2):
        if len(ip[i]) > 2:
            for j in range(len(ip[i])-2):
                aba = isAba(ip[i][j:(j+3)])
                if aba != ('_', '_'):
                    for k in range(1, len(ip), 2):
                        if (aba[1] + aba[0] + aba[1]) in ip[k]:
                            isSSL = True
    if isSSL:
        numOfSSL += 1

print "Part one: " + str(numOfTLS)
print "Part two: " + str(numOfSSL)

File: advall-haskell/d7.hs
import Data.List.Split
import Data.List

hasABBA :: String -> Bool
hasABBA s
    | (length s) < 4 = False
hasABBA (s:t:u:v:xs)
    | (s == v) && (t == u) && (s /= t) = True
    | otherwise = hasABBA (t:u:v:xs)

lstHasABBA :: [String] -> Bool
lstHasABBA [] = False
lstHasABBA (x:[]) = hasABBA x
lstHasABBA (x:y:xs) = (hasABBA x) || (lstHasABBA xs)

isTLS :: [String] -> Bool
isTLS s = (lstHasABBA s) && (not (lstHasABBA (tail s)))

splitInputLine :: String -> [String]
splitInputLine s = splitWhen isSqrBracket s

isSqrBracket :: Char -> Bool
isSqrBracket c = (c == '[') || (c == ']') 

solve1 :: [String] -> Int
solve1 xs = length (filter isTLS (map splitInputLine xs))

getABAs :: String -> [String]
getABAs s
    | (length s) < 3 = []
getABAs (x:y:z:xs)
    | (x == z) && (x /= y) = [[x] ++ [y] ++ [z]] ++ getABAs (y:z:xs)
    | otherwise = getABAs (y:z:xs)

lstGetABAs :: [String] -> [String]
lstGetABAs [] = []
lstGetABAs (x:[]) = getABAs x
lstGetABAs (x:y:xs) = (getABAs x) ++ (lstGetABAs xs)

lstGetBABs  :: [String] -> [String]
lstGetBABs (x:xs) = map makeBAB (lstGetABAs xs)

makeBAB :: String -> String
makeBAB (a:b:aa) = [b, a, b]

isSSL :: [String] -> Bool
isSSL s = (length ((lstGetABAs s) `intersect` (lstGetBABs s))) > 0

solve2 :: [String] -> Int
solve2 xs = length (filter isSSL (map splitInputLine xs))


main = do
    input <- readFile "d7_input.txt"
    let answer1 = solve1 (lines input)
    let answer2 = solve2 (lines input)
    putStrLn $ "Part one: " ++ show answer1
    putStrLn $ "Part two: " ++ show answer2